### ISSUE TITLE
Request render mode doesn't always work with 3D Tiles or Models

### ISSUE THREAD

**Issue by mramato** [CesiumGS/cesium#6898]
If I fly from home view to a tileset in request render mode (and the tileset isn't cached), the flight happens but nothing ever gets rendered until I manually force one.

Does Cesium automatically trigger a render on flight complete?  Shouldn't it?

CC @ggetz

**Comment by mramato** [comment]
I can make this happen all of the time in ion because the preview window has nothing else going on (no world imagery or terrain) so a tileset is the only thing that needs to load and Cesium is missing an internal render request somewhere.

**Comment by mramato** [comment]
As a workaround, I ended up adding a setTimeout in the flyTo `complete` callback I was using, (just a straight up requestRender didn't work, it had to be on several hundred ms delay)

**Comment by ggetz** [comment]
@mramato It doesn't trigger one by default. It could.

I believe nothing is rendered because the tileset is not selected by rendering, therefore nothing begins loading. So there may a more direct way to make sure the tileset begins loading in request render mode than the flyTo completing. I'll take a look soon.

**Comment by mramato** [comment]
Do we already trigger a render every time a ajax request completes?  If not, could this be something where it's fetching content but doesn't render the content once it's fetched?

Otherwise, I'll leave it to your expertise.  Thanks.

**Comment by mramato** [comment]
We should really try to fix this for the next release. `requestRender` bugs on our end will make it hard for end users to use the feature and inserting random render calls in user code to work around it kind of negates the point.

**Comment by hpinkos** [comment]
I ran into another issue where the tileset doesn't clear from the globe on `scene.primitives.remove`.  calling `scene.requestRender` after `remove` didn't work either, I had to call `scene.forceRender`.

**Comment by mramato** [comment]
I'm running into similar problems with 3D model loading.

**Comment by mramato** [comment]
Here's a trivial way to reproduce this with Sandcastle [direct link](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=dVNNc9owEP0rGi6IjiNDaC6UME3JpD2UNgO0J1+EvWANsuRKMgnt5L93LWFqPqrxwdLue/v2a8cN2Ql4AUPuiYIXMgUrqoL99G+0m/rrVCvHhQLTjcifRBE8Bn5VYN0cVAZmpjMYEWcqSNRb70OiErVD4gKfJfIeOGsvydZGF5+lW9MDUWXkiHQZi4/fQ1naeMGLUsIjdzz2NDYOLDMht0tTpdvz+822SNlGrrpR4LUpKBQVkmP+drB4vhl3RryOGm1Lw5Vda1NYBty6b9q4/Ee51E/iFbInwwugB88pNw7/uBr6VB5hYwAsHUTkNiKDPp5e71gFC24pCtCVo+tKpU5oRWivqWFbGyuNKIQTO7CMZxn1ImuSC8eTytMQ55gWWnm2f0ZhwgJzOahrgT0nNqhueoCtdKUyoTaLMgeDMXgmKnukbvwxvvOT0nTUF/E9KyrpRCn3n/bPWigXxLNWnaPrYQJf1B68f+WlvSb9JnyOuSH4JH7OnJ7jM7aP3gyGfdY/R5XCpfn/Mbd3lxAchg0gxJB3ZMju2tZ2J1IcC8OZ1Hr74OiVZL4Ewc+1gnnNSQ8pREFVFCIdE/VT8xbGqP7tRJ2xdXsJk2D/KIoSJ7PeGYqL4gB3hGO54hUuADiWWlvDxnEDGmdiR0R2n3TOFjnpkFRya9GyrqRciN+QdCbjGP1PYFJ7vd93YCTf1y75YPI1PDLGxjFeL1FOa7nipsX4Fw): 

```js
var viewer = new Cesium.Viewer('cesiumContainer', {
    requestRenderMode: true
});

var model = Cesium.Model.fromGltf({
    url: '../../../../Apps/SampleData/models/CesiumMilkTruck/CesiumMilkTruck-kmc.glb',
    scene: viewer.scene,
    modelMatrix: Cesium.Transforms.eastNorthUpToFixedFrame(Cesium.Cartesian3.fromDegrees(1, 2, 10000))
});

setTimeout(function () {
    viewer.scene.primitives.add(model);
    viewer.scene.requestRender();

    model.readyPromise.then(function () {
        var r = model.boundingSphere.radius;

        var center = Cesium.Matrix4.multiplyByPoint(model.modelMatrix, model.boundingSphere.center, new Cesium.Cartesian3());
        var heading = Cesium.Math.toRadians(-130.0);
        var pitch = Cesium.Math.toRadians(-25.0);
        var range = r * 3.5;
        viewer.scene.camera.lookAt(center, new Cesium.HeadingPitchRange(heading, pitch, range));
    });
}, 1000);
```

It doesn't happen every time (because race condition), but most of the time the model.readyPromise will never resolve until the user interfaces with the window or resizes the screen (even though we call requestRender after adding the primitive)

@ggetz does this help narrow down the issue?

**Comment by mramato** [comment]
Also reported in #7489

**Comment by KermMartian** [comment]
Apologies for bumping this (via my #7489). I wanted to check if there was a target release for resolving this, or if there was anything I could do to help with its resolution. Thanks!

**Comment by hpinkos** [comment]
@KermMartian we don't have a specific release target, but we're hoping to look into this soon.  There are a few other things in-progress we need to finish first though.

If you're interested in looking into a fix, we would be happy to review a pull request!  @ggetz should be able to give some suggestions for where to look for the problem

**Comment by mramato** [comment]
@likangning93 @lilleyse @ggetz this is the issue I was mentioning yesterday.

**Comment by KermMartian** [comment]
I'm enthused if this is going to be addressed soon. :) I'm not pleased with our current solution, which is to not use the request render mode.

**Comment by mramato** [comment]
A lot has changed since this issue was written.  Can someone verify that there are still actual problems here?

@KermMartian one workaround I've used successfully is to just request a render at least once a second with the below code:

```js
function requestRenderBug() {
    if (!viewer.isDestroyed()) {
        viewer.scene.requestRender();
        setTimeout(requestRenderBug, 1000);
    }
}
setTimeout(requestRenderBug, 1000);
```

Gets you most of the benefit while avoiding the bug.

**Comment by mramato** [comment]
I just ran into this using the latest Cesium, so definitely still a problem.

**Comment by lilleyse** [comment]
The problem is that `Model` requires at least one update before it becomes ready (and possibly more if `asynchronous` is true since the `JobScheduler` may take a few frames before it becomes ready) but update is never called until a user action occurs that triggers `requestRender`.

One approach is for scene to keep a list of primitives that still need updating request render until the list is empty. Another is for scene to listen to model load events like it does for RequestScheduler, TaskProcessor, and Globe. Neither approach is super well defined at the moment.

**Comment by ebogo1** [comment]
> A lot has changed since this issue was written. Can someone verify that there are still actual problems here?

I can still reproduce the bug with `model.readyPromise` not resolving until a render is forced by something like resizing the browser window - [updated Sandcastle](https://sandcastle.cesium.com/#c=dVNNc9owEP0rGi6IjitDaC6UMG3JpD2UlgHaky/CXrAGWXIlmYR28t+7ljA1H9VJ2t339u2H9tyQvYBnMOSBKHgmU7CiKthPb6Pd1D+nWjkuFJhuRP4kiuAx8KsC6xagMjAzncGIOFNBol577xOVqD0SF2iWyHvkrKMk2xhdfJZuQ49ElZEj0mUsXvKilPDIHY890MYBNxNytzJVurt8s61cd6PAYlNQKCGUwvzr6PFcM+6MeBk1SlaGK7vRprAMuHXftHH5j3Kln8QLZE+GF0CPkVNuHN64Gnrhj7A1AJYOInIXkUEfT693qtmCW4kCdOXoplKpE1oR2ms61tbGSiMK4cQeLONZRr3ImuQq8KzPNOQ5lYVenh3mKExYYC4HdSux58Rx1CMOsLWuVCbUdlnmYDAHz0RlT9RNPOZ3fi+a+fkmvmNFJZ0o5eHTYa6FckE8a/U5up0m8EXtNfvXXtprym/S51gbgs/y58zpBZpxfPTtYNhn/UtUKVya/x9zd38NwWXYAkIMeUOG7L7tbU8ixbUwnEmtdx8dvVHMlyB4XitY1Jz0WEIUVEUh06lQvzWvYY3qayfqjK07SJgE/wdRlLiZ9Q+h+D8c4P/g2K54jcsPjqXW1rBx3IDGmdgTkT0knYtvm3RIKrm16NlUUi7Fb0g6k3GM8Wcwqb3e73swkh/qkHww+RqMjLFxjM9rlNNarrlpMf4F). This is happening in sandcastle.cesium.com on CesiumJS 1.88. 

But, I cannot reproduce this issue locally, even on older versions of CesiumJS, both built and unbuilt. Even if I host a .glb to load on a local server as opposed to linking to it with a relative filepath, I still cannot reproduce locally. Has this ever been an issue in local development?

**Comment by ephjos** [comment]
I just ran into this issue (or something very similar) while developing locally. I was able use what @mramato [did above](https://github.com/CesiumGS/cesium/issues/6898#issuecomment-411528465) and resolve the issue. I noticed that the issue was only happening on Chrome (Brave and Chromium), but Firefox updates immediately without issue. I am able to reproduce this behavior with the sandcastle.

1. Open @ebogo1's [updated Sandcastle](https://sandcastle.cesium.com/#c=dVNNc9owEP0rGi6IjitDaC6UMG3JpD2UlgHaky/CXrAGWXIlmYR28t+7ljA1H9VJ2t339u2H9tyQvYBnMOSBKHgmU7CiKthPb6Pd1D+nWjkuFJhuRP4kiuAx8KsC6xagMjAzncGIOFNBol577xOVqD0SF2iWyHvkrKMk2xhdfJZuQ49ElZEj0mUsXvKilPDIHY890MYBNxNytzJVurt8s61cd6PAYlNQKCGUwvzr6PFcM+6MeBk1SlaGK7vRprAMuHXftHH5j3Kln8QLZE+GF0CPkVNuHN64Gnrhj7A1AJYOInIXkUEfT693qtmCW4kCdOXoplKpE1oR2ms61tbGSiMK4cQeLONZRr3ImuQq8KzPNOQ5lYVenh3mKExYYC4HdSux58Rx1CMOsLWuVCbUdlnmYDAHz0RlT9RNPOZ3fi+a+fkmvmNFJZ0o5eHTYa6FckE8a/U5up0m8EXtNfvXXtprym/S51gbgs/y58zpBZpxfPTtYNhn/UtUKVya/x9zd38NwWXYAkIMeUOG7L7tbU8ixbUwnEmtdx8dvVHMlyB4XitY1Jz0WEIUVEUh06lQvzWvYY3qayfqjK07SJgE/wdRlLiZ9Q+h+D8c4P/g2K54jcsPjqXW1rBx3IDGmdgTkT0knYtvm3RIKrm16NlUUi7Fb0g6k3GM8Wcwqb3e73swkh/qkHww+RqMjLFxjM9rlNNarrlpMf4F)
2. Allow the page to load
3. Click Run in the top-left
4. The viewer should reload

On Chrome the zoomed out earth view should be shown:
![sel-220217-1432-32](https://user-images.githubusercontent.com/35752047/154556891-4d694735-0df3-4e19-b04c-523ab0fcebd9.png)
If the user changes the size of the window or otherwise interacts with the viewer, the milk truck is shown.

On Firefox, the milk truck is shown immediately.

### Versions

- Brave Browser 97.1.34.81
- Chromium 97.0.4692.99 Arch Linux
- Mozilla Firefox 96.0.2
- Linux 5.16.2-arch1-1 x86_64 GNU/Linux

**Comment by lilleyse** [comment]
https://github.com/CesiumGS/cesium/issues/6898#issuecomment-991353207 describes why it's happening. We haven't done anything to fix it yet.

**Comment by mramato** [comment]
@ptrgags @lilleyse Did the glTF/3D Tiles next refactor happen to address this issue?

**Comment by lilleyse** [comment]
No, it was not addressed as part of the refactor.

I don't think it's too hard to implement though:

* Call `scene.requestRender()` when a tile is moved to the processing queue to ensure at least one `update` is called
* Call `scene.requestRender()` in [`processTiles`](https://github.com/CesiumGS/cesium/blob/70b1091605a842fb43168ae5139e13fadcb6fadd/packages/engine/Source/Scene/Cesium3DTileset.js#L2494-L2502) if length is greater than 0

Additionally so that individual models work:

* Call `scene.requestRender()` in [`processLoader`](https://github.com/CesiumGS/cesium/blob/70b1091605a842fb43168ae5139e13fadcb6fadd/packages/engine/Source/Scene/Model/Model.js#L1828)

One problem is that `scene` is not really accessible to either of these. So some sort of `requestRender` function might need to be added to `FrameState`.

**Comment by jhndrx** [comment]
> Additionally so that individual models work:
> 
> * Call `scene.requestRender()` in [`processLoader`](https://github.com/CesiumGS/cesium/blob/70b1091605a842fb43168ae5139e13fadcb6fadd/packages/engine/Source/Scene/Model/Model.js#L1828)
> 
> One problem is that `scene` is not really accessible to either of these. So some sort of `requestRender` function might need to be added to `FrameState`.

Request render mode has been causing me issues with models not loading until the user performs an action that forces a render. Like mentioned earlier in this thread, it can take multiple render calls before a model is ready and shows. In the `processLoader` method, I added the lambda `frameState.afterRender.push(() => true);` and it appears to make models load correctly. Of course, its rendering at its full rate until the model is ready but after they show, the rendering stops as desired.

**Comment by ggetz** [comment]
After https://github.com/CesiumGS/cesium/pull/11059 and adding `requestRender` to the loaders, I can no longer replicate. I did update the [sandcastle example to use the newer APIs](https://sandcastle.cesium.com/#c=bZJNc5swEIb/ioaL8Qwj4vbmEk8Tp+klbmdq2hOXNVpsTYREVxIO7fi/V4DJxHF12s+Hd3dpgVgr8YjEbpnGI1ujlb7mv4ZYPCsHd220A6mRZgn7W2gWHuFvj9b9QC2QNkbgkjnyWOjT/FOhC90GcB3CKnDhCNJN5L5W8YpM/VW56s52uozPTE9qyWacp1uoG4UP4CAdGDYdmzdSPefky+f3Pt+r3SwZKbZEHdSMU/HBO2cG1gYcyZflJCcn0LYyVFuOYN03Q+7ws8nNo3xB8UhQY3yuXAO5YIH+OKh/wD0h2niRsA8JW9yEN5+/jm/R5bJG411ceV06aTSL59Py3mrjDclaOtmi5SBEPIjsIVeFFyuPx++8jhWyILovLWrXYwbjSVoXGul/Eq7oZRiVgFeqy8298VpIvd82ByQcJfHdRTD8CEx4gh67ZDfsNEkejNO4kd6MkiizrlO4GvOfZd2EJffHjsOpHYZTQ1hsugt3RMdLa/u2LJ2aMiFbJsVtEb37GYuIlQqsDZnKK7WVf7CIVlka6i/alIFe9/cWSUHXlxwWq6cxyDnP0uBedzlj1A7oDfEf). 

Closing this as fixed.

**Comment by mramato** [comment]
@ggetz I'm confused by the new example. The doc for request render mode includes the following:

> If true, rendering a frame will only occur when needed as determined by changes within the scene.

I could be wrong, but historically adding a primitive to the scene was considered an internally detected "Change within the scene", but in your example it seems you have to call `requestRender` after adding the primitive. Did something change? At the very least, the documentation probably need to be a little clearer hear around when a render needs to be specifically requested.
